<a href="https://colab.research.google.com/github/AmitKumar-001/Learn-Data-Analysis/blob/main/Data_Cleaning_Melbourne_house_pricing.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [8]:
#import library
import numpy as np
import pandas as pd

In [67]:
#load & read csv
df = pd.read_csv("melbourne.csv")
df.head()

Suburb             Address  Rooms Type      Price Method SellerG  \
0  Abbotsford       68 Studley St      2    h        NaN     SS  Jellis   
1  Abbotsford        85 Turner St      2    h  1480000.0      S  Biggin   
2  Abbotsford     25 Bloomburg St      2    h  1035000.0      S  Biggin   
3  Abbotsford  18/659 Victoria St      3    u        NaN     VB  Rounds   
4  Abbotsford        5 Charles St      3    h  1465000.0     SP  Biggin   

         Date  Distance  Postcode  ...  Bathroom  Car  Landsize  BuildingArea  \
0  03-09-2016       2.5    3067.0  ...       1.0  1.0     126.0           NaN   
1  03-12-2016       2.5    3067.0  ...       1.0  1.0     202.0           NaN   
2  04-02-2016       2.5    3067.0  ...       1.0  0.0     156.0          79.0   
3  04-02-2016       2.5    3067.0  ...       2.0  1.0       0.0           NaN   
4  04-03-2017       2.5    3067.0  ...       2.0  0.0     134.0         150.0   

   YearBuilt  CouncilArea Lattitude  Longtitude             Regionname  \
0        NaN        Yarra  -37.8014    144.9958  Northern Metropolitan   
1        NaN        Yarra  -37.7996    144.9984  Northern Metropolitan   
2     1900.0        Yarra  -37.8079    144.9934  Northern Metropolitan   
3        NaN        Yarra  -37.8114    145.0116  Northern Metropolitan   
4     1900.0        Yarra  -37.8093    144.9944  Northern Metropolitan   

  Propertycount  
0        4019.0  
1        4019.0  
2        4019.0  
3        4019.0  
4        4019.0  

[5 rows x 21 columns]

In [68]:
#view overview of the data (here 23547 rows and 21 Columns)
print(df.shape)
#don't use like print(df.shape()) or print(df.info)
print(df.info())

(23547, 21)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 23547 entries, 0 to 23546
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Suburb         23547 non-null  object 
 1   Address        23547 non-null  object 
 2   Rooms          23547 non-null  int64  
 3   Type           23547 non-null  object 
 4   Price          18396 non-null  float64
 5   Method         23547 non-null  object 
 6   SellerG        23547 non-null  object 
 7   Date           23547 non-null  object 
 8   Distance       23546 non-null  float64
 9   Postcode       23546 non-null  float64
 10  Bedroom2       19066 non-null  float64
 11  Bathroom       19063 non-null  float64
 12  Car            18921 non-null  float64
 13  Landsize       17410 non-null  float64
 14  BuildingArea   10018 non-null  float64
 15  YearBuilt      11540 non-null  float64
 16  CouncilArea    15656 non-null  object 
 17  Lattitude      19243 non-null  float64

In [69]:
#check for null >> .isnull() return true false for null
df.isnull()

Suburb  Address  Rooms   Type  Price  Method  SellerG   Date  Distance  \
0       False    False  False  False   True   False    False  False     False   
1       False    False  False  False  False   False    False  False     False   
2       False    False  False  False  False   False    False  False     False   
3       False    False  False  False   True   False    False  False     False   
4       False    False  False  False  False   False    False  False     False   
...       ...      ...    ...    ...    ...     ...      ...    ...       ...   
23542   False    False  False  False   True   False    False  False     False   
23543   False    False  False  False   True   False    False  False     False   
23544   False    False  False  False  False   False    False  False     False   
23545   False    False  False  False  False   False    False  False     False   
23546   False    False  False  False  False   False    False  False     False   

       Postcode  ...  Bathroom    Car  Landsize  BuildingArea  YearBuilt  \
0         False  ...     False  False     False          True       True   
1         False  ...     False  False     False          True       True   
2         False  ...     False  False     False         False      False   
3         False  ...     False  False     False          True       True   
4         False  ...     False  False     False         False      False   
...         ...  ...       ...    ...       ...           ...        ...   
23542     False  ...     False  False     False         False      False   
23543     False  ...      True   True      True          True       True   
23544     False  ...     False  False      True          True       True   
23545     False  ...     False  False     False         False      False   
23546     False  ...     False  False      True         False      False   

       CouncilArea  Lattitude  Longtitude  Regionname  Propertycount  
0            False      False       False       False          False  
1            False      False       False       False          False  
2            False      False       False       False          False  
3            False      False       False       False          False  
4            False      False       False       False          False  
...            ...        ...         ...         ...            ...  
23542         True      False       False       False          False  
23543         True      False       False       False          False  
23544         True      False       False       False          False  
23545         True      False       False       False          False  
23546         True      False       False       False          False  

[23547 rows x 21 columns]

In [34]:
#count missing values in columns by using df.isnull().sum()
df.isnull().sum()

Suburb               0
Address              0
Rooms                0
Type                 0
Price             5151
Method               0
SellerG              0
Date                 0
Distance             1
Postcode             1
Bedroom2          4481
Bathroom          4484
Car               4626
Landsize          6137
BuildingArea     13529
YearBuilt        12007
CouncilArea       7891
Lattitude         4304
Longtitude        4304
Regionname           1
Propertycount        1
dtype: int64

In [39]:
#now check columns having at least one missing value
df.isnull().any()

#above is equivalent to axis = 0 (by default, any() operates on columns)
#df.isnull().any(axis=0) #axis = 0 for columns
#df.isnull().any(axis=1) #axis = 1 for Rows

Suburb           False
Address          False
Rooms            False
Type             False
Price             True
Method           False
SellerG          False
Date             False
Distance          True
Postcode          True
Bedroom2          True
Bathroom          True
Car               True
Landsize          True
BuildingArea      True
YearBuilt         True
CouncilArea       True
Lattitude         True
Longtitude        True
Regionname        True
Propertycount     True
dtype: bool

In [42]:
#rows having all missing Values
df.isnull().all(axis=1)

0        False
1        False
2        False
3        False
4        False
         ...  
23542    False
23543    False
23544    False
23545    False
23546    False
Length: 23547, dtype: bool

In [43]:
#column having all missing Values
df.isnull().all(axis=0)

Suburb           False
Address          False
Rooms            False
Type             False
Price            False
Method           False
SellerG          False
Date             False
Distance         False
Postcode         False
Bedroom2         False
Bathroom         False
Car              False
Landsize         False
BuildingArea     False
YearBuilt        False
CouncilArea      False
Lattitude        False
Longtitude       False
Regionname       False
Propertycount    False
dtype: bool

In [45]:
#check how many rows having all missing values do sum
df.isnull().all(axis=1).sum() #for row axis =1

np.int64(0)

In [46]:
#check how many column having all missing values do sum
df.isnull().all(axis=0).sum() #for column axis =0

np.int64(0)

In [70]:
#count missing values in each rows
df.isnull().sum(axis=1)

0        3
1        2
2        0
3        3
4        0
        ..
23542    2
23543    8
23544    4
23545    1
23546    2
Length: 23547, dtype: int64

In [63]:
#count missing values in each column
df.isnull().sum(axis=0)

Suburb              0
Address             0
Rooms               0
Type                0
Price            5151
Method              0
SellerG             0
Date                0
Distance            1
Postcode            1
Bedroom2         4481
Bathroom         4484
Car              4626
Landsize         6137
Lattitude        4304
Longtitude       4304
Regionname          1
Propertycount       1
dtype: int64

In [71]:
#to Impute: Techniques for prediction: To predict and impute missing values, use statistical models such as kNN, SVM, and others.
#sum up the missing vlues (Column Wise)
round(100*(df.isnull().sum()/len(df.index)), 2) #missing percetage

Suburb            0.00
Address           0.00
Rooms             0.00
Type              0.00
Price            21.88
Method            0.00
SellerG           0.00
Date              0.00
Distance          0.00
Postcode          0.00
Bedroom2         19.03
Bathroom         19.04
Car              19.65
Landsize         26.06
BuildingArea     57.46
YearBuilt        50.99
CouncilArea      33.51
Lattitude        18.28
Longtitude       18.28
Regionname        0.00
Propertycount     0.00
dtype: float64

In [72]:
#removing the columns whose are mostly missing values like more than 30%
df.drop(['BuildingArea', 'YearBuilt', 'CouncilArea'], axis=1, inplace=True) #the columns has removed

In [74]:
df.shape #earliear it was 21 column now it is only 21

(23547, 18)

In [75]:
#now no anyone has more than 30% missing values
round(100*(df.isnull().sum()/len(df.index)), 2)

Suburb            0.00
Address           0.00
Rooms             0.00
Type              0.00
Price            21.88
Method            0.00
SellerG           0.00
Date              0.00
Distance          0.00
Postcode          0.00
Bedroom2         19.03
Bathroom         19.04
Car              19.65
Landsize         26.06
Lattitude        18.28
Longtitude       18.28
Regionname        0.00
Propertycount     0.00
dtype: float64

In [76]:
#check if anyone has more than 5 missing values
df[df.isnull().sum(axis=1) > 5]

Suburb              Address  Rooms Type      Price Method  \
15            Abbotsford     217 Langridge St      3    h  1000000.0      S   
16            Abbotsford      18a Mollison St      2    t   745000.0      S   
19            Abbotsford  403/609 Victoria St      2    u   542000.0      S   
21            Abbotsford    25/84 Trenerry Cr      2    u   760000.0     SP   
22            Abbotsford    106/119 Turner St      1    u   481000.0     SP   
...                  ...                  ...    ...  ...        ...    ...   
23495  Templestowe Lower  1/207 Manningham Rd      2    u   550000.0     PI   
23499          Thornbury      1/128 Dundas St      2    t   770000.0     PI   
23502          Thornbury        111 Pender St      2    t   858000.0      S   
23508             Toorak   21/1059 Malvern Rd      2    u   720000.0     VB   
23531           Werribee         91 Latham St      4    h   540000.0     PI   

            SellerG        Date  Distance  Postcode  Bedroom2  Bathroom  Car  \
15           Jellis  08-10-2016       2.5    3067.0       NaN       NaN  NaN   
16           Jellis  08-10-2016       2.5    3067.0       NaN       NaN  NaN   
19           Dingle  08-10-2016       2.5    3067.0       NaN       NaN  NaN   
21           Biggin  10-12-2016       2.5    3067.0       NaN       NaN  NaN   
22     Purplebricks  10-12-2016       2.5    3067.0       NaN       NaN  NaN   
...             ...         ...       ...       ...       ...       ...  ...   
23495         Barry  26-08-2017      12.4    3107.0       NaN       NaN  NaN   
23499       McGrath  26-08-2017       7.0    3071.0       NaN       NaN  NaN   
23502        Jellis  26-08-2017       7.0    3071.0       NaN       NaN  NaN   
23508        Beller  26-08-2017       4.1    3142.0       NaN       NaN  NaN   
23531       Triwest  26-08-2017      14.7    3030.0       NaN       NaN  NaN   

       Landsize  Lattitude  Longtitude             Regionname  Propertycount  
15          NaN        NaN         NaN  Northern Metropolitan         4019.0  
16          NaN        NaN         NaN  Northern Metropolitan         4019.0  
19          NaN        NaN         NaN  Northern Metropolitan         4019.0  
21          NaN        NaN         NaN  Northern Metropolitan         4019.0  
22          NaN        NaN         NaN  Northern Metropolitan         4019.0  
...         ...        ...         ...                    ...            ...  
23495       NaN        NaN         NaN   Eastern Metropolitan         5420.0  
23499       NaN        NaN         NaN  Northern Metropolitan         8870.0  
23502       NaN        NaN         NaN  Northern Metropolitan         8870.0  
23508       NaN        NaN         NaN  Southern Metropolitan         7217.0  
23531       NaN        NaN         NaN   Western Metropolitan        16166.0  

[4278 rows x 18 columns]

In [78]:
#count the number of rows having > 5 missing values
# use len(df.index) #need to deep dive in it with help of another sources
len(df[df.isnull().sum(axis=1) > 5].index)

4278

In [79]:
#4278 rows have more than 5 missing values
#Calculate the percentage
round(100*(len(df[df.isnull().sum(axis=1) > 5].index)/len(df.index)), 2)

18.17

In [80]:
#retaining the rows only having < 5 missing values
df = df[df.isnull().sum(axis=1) <= 5]

In [81]:
#now check summary agian
round(100*(df.isnull().sum()/len(df.index)), 2)

Suburb            0.00
Address           0.00
Rooms             0.00
Type              0.00
Price            21.71
Method            0.00
SellerG           0.00
Date              0.00
Distance          0.00
Postcode          0.00
Bedroom2          1.05
Bathroom          1.07
Car               1.81
Landsize          9.65
Lattitude         0.13
Longtitude        0.13
Regionname        0.00
Propertycount     0.00
dtype: float64

In [85]:
#removing missing price rows
df = df[~np.isnan(df['Price'])] #need to deep dive in it with external sources

In [86]:
#now view summary again
round(100*(df.isnull().sum()/len(df.index)), 2)

Suburb           0.00
Address          0.00
Rooms            0.00
Type             0.00
Price            0.00
Method           0.00
SellerG          0.00
Date             0.00
Distance         0.00
Postcode         0.00
Bedroom2         1.05
Bathroom         1.07
Car              1.76
Landsize         9.83
Lattitude        0.15
Longtitude       0.15
Regionname       0.00
Propertycount    0.00
dtype: float64

In [88]:
#focus more missing value like landsize
df['Landsize'].describe() #do deep dive in it

count     13603.000000
mean        558.116371
std        3987.326586
min           0.000000
25%         176.500000
50%         440.000000
75%         651.000000
max      433014.000000
Name: Landsize, dtype: float64

Here the minimum is 0, max is 433014, the mean is 558 and median (50%) is 440. There is a significant variation in the 25th and 75th percentile as well (176 to 651). Thus, imputing this with mean/median seems quite biased and so we should remove the missing value.

In [89]:
#removing missig values from landsize
df = df[~np.isnan(df['Landsize'])]

In [91]:
#now recheck % missing
round(100*(df.isnull().sum()/len(df.index)), 2)

Suburb           0.00
Address          0.00
Rooms            0.00
Type             0.00
Price            0.00
Method           0.00
SellerG          0.00
Date             0.00
Distance         0.00
Postcode         0.00
Bedroom2         0.00
Bathroom         0.01
Car              0.46
Landsize         0.00
Lattitude        0.16
Longtitude       0.16
Regionname       0.00
Propertycount    0.00
dtype: float64

In [95]:
#now going to impute the missing value rather than removing
#rows having lattitude and longitude missing
df[np.isnan(df['Lattitude'])]

Suburb                Address  Rooms Type      Price  \
2572              Burwood           23 Monica St      3    h   990000.0   
3257         Clifton Hill  3/268 Alexandra Pde E      1    u   363000.0   
4485            Footscray          483 Barkly St      3    t   781000.0   
5170         Hampton East           7 Seafoam St      4    t  1185000.0   
10745  Williamstown North         4/9 Adeline St      1    u   355000.0   
13223           Melbourne   1913/228 Abeckett St      3    u  1175000.0   
14008            Brooklyn          9 Richards Ct      3    h   750000.0   
14132     North Melbourne   13/201 Abbotsford St      2    t   755000.0   
14139      Oakleigh South            4 Druitt St      4    h  1205500.0   
14142      Oakleigh South        298 Warrigal Rd      3    h   799999.0   
14143      Oakleigh South            11 Yarra Ct      5    h  1200000.0   
14976            Essendon        9 Washington St      3    h  1520000.0   
15278              Seddon          60 Station Rd      4    h  1500000.0   
15561             Croydon       3 Silvergrass Ct      3    h   630000.0   
16358          Kensington     201/102 Rankins Rd      2    u   876000.0   
16564          Strathmore          48 Lebanon St      3    h   950000.0   
17616          Kensington     109/102 Rankins Rd      1    u   410000.0   
17620         Keysborough          19 Denmark Rd      5    h  1100000.0   
18141               Lalor           83 Rotino Cr      3    t   463000.0   
18614           Mickleham        17 Primavera Dr      4    h   610000.0   
18793             Wollert      13 Strathalbyn Ch      4    h   631000.0   
18995           Greenvale         40 Frontier Av      3    t   470000.0   

      Method        SellerG        Date  Distance  Postcode  Bedroom2  \
2572      VB      Fletchers  17-09-2016      11.7    3125.0       3.0   
3257       S  hockingstuart  27-06-2016       3.4    3068.0       1.0   
4485       S            Jas  27-11-2016       6.4    3011.0       3.0   
5170       S             RT  28-05-2016      14.5    3188.0       4.0   
10745      S        Sweeney  27-11-2016       8.9    3016.0       1.0   
13223     PI           Icon  29-04-2017       2.8    3000.0       3.0   
14008      S  hockingstuart  20-05-2017      10.9    3012.0       3.0   
14132     PI         Nelson  29-04-2017       2.3    3051.0       2.0   
14139      S       Woodards  22-04-2017      14.7    3167.0       4.0   
14142      S       Woodards  29-04-2017      14.7    3167.0       3.0   
14143     PI         Buxton  29-04-2017      14.7    3167.0       5.0   
14976      S           Brad  03-06-2017       7.5    3040.0       3.0   
15278     SP        Sweeney  03-06-2017       5.1    3011.0       4.0   
15561      S        McGrath  17-06-2017      23.0    3136.0       3.0   
16358      S        Rendina  24-06-2017       3.4    3031.0       2.0   
16564     PI      Considine  24-06-2017       8.2    3041.0       3.0   
17616     PI        Rendina  08-07-2017       3.4    3031.0       1.0   
17620     PI       VicHomes  08-07-2017      25.2    3173.0       5.0   
18141      S            HAR  12-08-2017      16.3    3075.0       3.0   
18614      S            Ray  15-07-2017      20.6    3064.0       4.0   
18793     PI  hockingstuart  15-07-2017      25.5    3750.0       4.0   
18995     SP          Barry  22-07-2017      20.4    3059.0       3.0   

       Bathroom  Car  Landsize  Lattitude  Longtitude  \
2572        2.0  2.0     263.0        NaN         NaN   
3257        1.0  1.0       0.0        NaN         NaN   
4485        2.0  2.0      98.0        NaN         NaN   
5170        3.0  1.0     300.0        NaN         NaN   
10745       1.0  1.0      46.0        NaN         NaN   
13223       3.0  2.0       0.0        NaN         NaN   
14008       1.0  2.0     667.0        NaN         NaN   
14132       1.0  1.0    1537.0        NaN         NaN   
14139       2.0  2.0     553.0        NaN         NaN   
14142       2.0  4.0     590.0        NaN  

In [99]:
#summary of the rows having missing lattitude and longitude so if possible we can impute the misssig value
#don't use like >> df[np.isnan(df['Lattitude'])].describe()
df.loc[:,['Lattitude','Longtitude']].describe() #deep dive to .loc[:,]

Lattitude    Longtitude
count  13581.000000  13581.000000
mean     -37.809204    144.995221
std        0.079257      0.103913
min      -38.182550    144.431810
25%      -37.856820    144.929600
50%      -37.802360    145.000100
75%      -37.756400    145.058320
max      -37.408530    145.526350

Here, there is no big difference between (max, min) or (mean, median) or variation from 25th to 75th percentile. (not comparing between Lattitude and Longtitude). We can impute the missing values by mean value respectively.

In [100]:
#imputing Lattitude and Longitude by mean values
df.loc[np.isnan(df['Lattitude']), ['Lattitude']] = df['Lattitude'].mean()
df.loc[np.isnan(df['Longtitude']), ['Longtitude']] = df['Longtitude'].mean()
#need to deep dive

In [101]:
#now check missing percentage
round(100*(df.isnull().sum()/len(df.index)), 2)

Suburb           0.00
Address          0.00
Rooms            0.00
Type             0.00
Price            0.00
Method           0.00
SellerG          0.00
Date             0.00
Distance         0.00
Postcode         0.00
Bedroom2         0.00
Bathroom         0.01
Car              0.46
Landsize         0.00
Lattitude        0.00
Longtitude       0.00
Regionname       0.00
Propertycount    0.00
dtype: float64

Still Bathroom and Car has missing values so now let either impute or remove

In [102]:
#focus on Bathroom and Car
df.loc[:,['Bathroom', 'Car']].describe()

Bathroom           Car
count  13602.000000  13540.000000
mean       1.534921      1.610414
std        0.691834      0.962244
min        0.000000      0.000000
25%        1.000000      1.000000
50%        1.000000      2.000000
75%        2.000000      2.000000
max        8.000000     10.000000

These two are integer type variables, and thus have values 0, 1, 2, 5 etc. We can't impute the missing values by the mean or the median (like 1.53 bathrooms doesn't make sense!).
Thus, we need to impute them by the mode (the most common occurring value)

In [105]:
#converting to type 'category'
df['Car'] = df['Car'].astype('category') #need to deep dive

#displaying frequencies of each category
df['Car'].value_counts()

Car
2.0     5606
1.0     5515
0.0     1026
3.0      748
4.0      507
5.0       63
6.0       54
8.0        9
7.0        8
10.0       3
9.0        1
Name: count, dtype: int64

here mode is 2 as it is most common (5606) so we will replace the missing value by 2

In [106]:
#replace missing value by 2
df.loc[pd.isnull(df['Car']),['Car']] = 2

In [107]:
#now check the missing percentage
round(100*(df.isnull().sum()/len(df.index)), 2)

Suburb           0.00
Address          0.00
Rooms            0.00
Type             0.00
Price            0.00
Method           0.00
SellerG          0.00
Date             0.00
Distance         0.00
Postcode         0.00
Bedroom2         0.00
Bathroom         0.01
Car              0.00
Landsize         0.00
Lattitude        0.00
Longtitude       0.00
Regionname       0.00
Propertycount    0.00
dtype: float64

now only bathroom has missing value so now let work on it either impute or remove

In [109]:
#Similarly doing for Bathroom as above did for car
#converting to type 'category'
df['Bathroom'] = df['Bathroom'].astype('category') #need to deep dive

#displaying frequencies of each category
df['Bathroom'].value_counts()

Bathroom
1.0    7517
2.0    4987
3.0     921
4.0     106
0.0      34
5.0      28
6.0       5
7.0       2
8.0       2
Name: count, dtype: int64

In [110]:
#here mode of bathroom is 1 (7517 the highest) so we can impute the missing value by 1
df.loc[pd.isnull(df['Bathroom']),['Bathroom']] = 2

In [111]:
#now check the missing percentage
round(100*(df.isnull().sum()/len(df.index)), 2)

Suburb           0.0
Address          0.0
Rooms            0.0
Type             0.0
Price            0.0
Method           0.0
SellerG          0.0
Date             0.0
Distance         0.0
Postcode         0.0
Bedroom2         0.0
Bathroom         0.0
Car              0.0
Landsize         0.0
Lattitude        0.0
Longtitude       0.0
Regionname       0.0
Propertycount    0.0
dtype: float64

Now none of them are missing

In [112]:
#now no missing values
#review the data shape after cleaning the data
df.shape

(13603, 18)

In [147]:
#we started with (23547, 21) shape and now we have (13603, 18)
#fraction of rows lost
print("fraction of rows have",round(100*len(df.index)/23547,2),"%")
print("fraction of rows lost",100-round(100*len(df.index)/23547,2),"%")
print("fraction of column have",round(100*len(df.columns)/21,2),"%")
#print("fraction of column have",100-round(100*len(df.columns)/21,2),"%") #this doesn't round the value it gives 14.290000000000006 %
print("fraction of column have",round(100-100*len(df.columns)/21,2),"%")

fraction of rows have 57.77 %
fraction of rows lost 42.23 %
fraction of column have 85.71 %
fraction of column have 14.29 %


Now the data has cleaned and we can go for analysis part